sale_coupon_multi_use/models/sale_coupon.py

```diff
@@ -0,0 +1,154 @@
+            "sale_order_line_id": line.id,
+        }
+
+    def _create_consumption_line(self, amount_consumed, sale_order):
+        vals = self._prepare_consumption_line(amount_consumed, sale_order)
+        return self.env["sale.coupon.consumption_line"].create(vals)
+
+    def _consume_line(self, discount, sale_order):
+        self.ensure_one()
+        amount_consumed = min(discount, self.discount_fixed_amount_delta)
+        consumption_line = self._create_consumption_line(amount_consumed, sale_order)
+        consumption_line._normalize_discount()
+        return consumption_line
+
+    def _handle_multi_use(self, coupon_order_data):
+        self.ensure_one()
+        new_delta, discount = self._get_compared_discount_with_delta(coupon_order_data)
+        self._consume_line(discount, coupon_order_data["order"])
+        # Indicates whether coupon was fully consumed.
+        return new_delta <= 0
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """Extend to pick up coupon_multi_use value from program."""
+        for vals in vals_list:
+            multi_use = (
+                self.env["sale.coupon.program"]
+                .browse(vals.get("program_id"))
+                .coupon_multi_use
+            )
+            if "multi_use" not in vals:
+                vals["multi_use"] = multi_use
+        return super().create(vals_list)
+
+    def write(self, vals):
+        """Extend to manage multi_use coupons."""
+        if self._is_multi_use_triggered(vals):
+            coupon_order_data = self._context.get("coupon_order_data")
+            if coupon_order_data:
+                coupon_consumed = self._handle_multi_use(coupon_order_data)
+                if not coupon_consumed:
+                    # Makes it so coupon that coupon is not used up yet.
+                    del vals["state"]
+        return super().write(vals)
+
+
+class SaleCouponConsumptionLine(models.Model):
+    """Model that stores data for single coupon multiple uses."""
+
+    _name = "sale.coupon.consumption_line"
+    _description = "Sale Coupon Consumption Line"
+
+    coupon_id = fields.Many2one("sale.coupon", "Coupon", required=True, index=True)
+    # ondelete takes care of automatically removing consumption line,
+    # when discount line is removed on related sale order.
+    sale_order_line_id = fields.Many2one(
+        "sale.order.line", "Sale Order Line", required=True, ondelete="cascade"
+    )
+    amount = fields.Float()
+
+    def _normalize_discount(self):
+        """Adjust SOL discount to match consumed discount."""
+        self.ensure_one()
+        # Discount initially won't match, when standard functionality
+        # applies full discount from coupon. But because we split
+        # coupon amount, we want to apply maximum possible discount.
+        sol = self.sale_order_line_id
+        sol_discount = abs(sol.price_unit)
+        amount_to_adjust = sol_discount - self.amount
+        if amount_to_adjust > 0:
```

Comment:
Please use float_compare()

Reply:
And precision to use? The one from related product UOM?